plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/PasteDataImportFromClipboardCommandHandler.java

```diff
@@ -0,0 +1,146 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Sebastian Hollersbacher - initial API and implementation and/or initial documentation
+ *******************************************************************************/
+package org.eclipse.fordiac.ide.gef.nat;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.BiFunction;
+import java.util.function.Supplier;
+
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.emf.ecore.util.EcoreUtil;
+import org.eclipse.fordiac.ide.model.commands.create.AddNewImportCommand;
+import org.eclipse.fordiac.ide.model.helpers.ImportHelper;
+import org.eclipse.fordiac.ide.model.helpers.PackageNameHelper;
+import org.eclipse.fordiac.ide.model.libraryElement.LibraryElement;
+import org.eclipse.fordiac.ide.model.typelibrary.TypeEntry;
+import org.eclipse.fordiac.ide.model.typelibrary.TypeLibrary;
+import org.eclipse.fordiac.ide.ui.widget.ImportTransfer;
+import org.eclipse.fordiac.ide.ui.widget.NatTableColumn;
+import org.eclipse.fordiac.ide.ui.widget.NatTableColumnProvider;
+import org.eclipse.fordiac.ide.ui.widget.PasteDataFromClipboardCommandHandler;
+import org.eclipse.gef.commands.CommandStack;
+import org.eclipse.nebula.widgets.nattable.copy.command.PasteDataCommand;
+import org.eclipse.nebula.widgets.nattable.data.ListDataProvider;
+import org.eclipse.nebula.widgets.nattable.layer.DataLayer;
+import org.eclipse.nebula.widgets.nattable.selection.SelectionLayer;
+import org.eclipse.swt.dnd.Clipboard;
+import org.eclipse.swt.widgets.Display;
+
+public class PasteDataImportFromClipboardCommandHandler extends PasteDataFromClipboardCommandHandler {
+
+	private final BiFunction<TypeLibrary, String, TypeEntry> typeResolver;
+	private final Supplier<CommandStack> cmdstk;
+	private final Map<String, String> conflicts = new HashMap<>();
+	private final NatTableColumnProvider<? extends NatTableColumn> columnProvider;
+	private final List<? extends NatTableColumn> columns;
+
+	public PasteDataImportFromClipboardCommandHandler(final SelectionLayer selectionLayer,
+			final Supplier<CommandStack> cmdStack, final BiFunction<TypeLibrary, String, TypeEntry> typeResolver,
+			final NatTableColumnProvider<? extends NatTableColumn> columnProvider,
+			final List<? extends NatTableColumn> columns) {
+		super(selectionLayer);
+		this.cmdstk = cmdStack;
+		this.typeResolver = typeResolver;
+		this.columnProvider = columnProvider;
+		this.columns = columns;
+	}
+
+	@Override
+	protected boolean doCommand(final PasteDataCommand command) {
+		final LibraryElement rootElement = getRootElement();
+		if (rootElement != null) {
+			final var namespaces = Arrays.stream(getClipboardContent()).map(imp -> getImportNamespace(rootElement, imp))
+					.filter(Objects::nonNull).toList();
+			super.doCommand(command);
+			namespaces.forEach(namespace -> cmdstk.get().execute(new AddNewImportCommand(rootElement, namespace)));
```

Comment:
I believe you need to add the imports before pasting the type in the super method for it to be resolved correctly?

Reply:
Yes, you are correct. Thank you for pointing that out
Unfortunately, this introduces the bug, where the selection gets cleared after importing (for the StructEditingComposite).

Reply:
My guess would be that the problem is caused by a refresh when adding the import? I would suggest to change the handler to use a `CommandExecutor` instead of the `CommandStack` and add a flag to block the refresh in `StructEditingComposite.executeCommand(Command)`, similar to `AbstractSection.executeCommand(Command)`.